### PR TITLE
1293 Revision/Amendment committing workflow based on revision status

### PR DESCRIPTION
**Pull request by Sepehr-Sobhani**
[ZH CARD](https://app.zenhub.com/workspaces/climate-action-secretariat-60ca4121764d710011481ca2/issues/gh/bcgov/cas-cif/1293)

### CODE REVIEW

File: app/tests/unit/lib/theme/widgets/ChangeReasonWidget.test.tsx [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+describe("The ChangeReasonWidget", () => {
+  beforeEach(() => {
+    componentTestingHelper.reinit();
+  });
+
+  it("renders the change reason textbox widget with default value along with an action button", () => {
+    componentTestingHelper.loadQuery();
+    componentTestingHelper.renderComponent();
+
```
Comment (BCerki) [inline app/tests/unit/lib/theme/widgets/ChangeReasonWidget.test.tsx:62]: Should there be an assertion for the default value? Or that doesn't need to be in the test title?

File: app/tests/unit/lib/theme/widgets/ChangeReasonWidget.test.tsx [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    const newChangeReason = "Another reason for change";
+
+    act(() => {
+      fireEvent.change(commentTextBox, {
+        target: { value: "Another reason for change" },
+      });
+    });
+    expect(handleChangeReason).toHaveBeenCalledWith(newChangeReason);
+
+    componentTestingHelper.rerenderComponent(undefined, {
```
Comment (BCerki) [inline app/tests/unit/lib/theme/widgets/ChangeReasonWidget.test.tsx:84]: Cool, I've never seen `rerenderComponent` used

File: app/tests/unit/lib/theme/widgets/RevisionStatusWidget.test.tsx [hunk 3]
```diff
@@ -33,33 +23,53 @@ const mockQueryPayload = {
...
+      label: "Revision Status",
+      value: "Draft",
+      default: undefined,
+      enum: undefined,
+      type: "string",
       schema: {
         anyOf: [
-          { value: 1, enum: [1], type: "number", title: "Option 1" },
-          { value: 2, enum: [2], type: "number", title: "Option 2" },
+          { value: "Draft", enum: ["Draft"], type: "string", title: "Draft" },
```
Comment (BCerki) [inline app/tests/unit/lib/theme/widgets/RevisionStatusWidget.test.tsx:53]: Could this be a variable? It's used again later

File: app/tests/unit/lib/theme/widgets/RevisionStatusWidget.test.tsx [hunk 4]
```diff
@@ -118,4 +171,57 @@ describe("The RevisionStatusWidget", () => {
     expect(screen.queryByRole("combobox")).not.toBeInTheDocument();
     expect(screen.queryByText(/update/i)).not.toBeInTheDocument();
   });
+  it("calls commitProjectRevision when the select value is `Applied`", () => {
+    componentTestingHelper.loadQuery();
+    componentTestingHelper.renderComponent();
+    const newStatus = "Applied";
```
Comment (BCerki) [inline app/tests/unit/lib/theme/widgets/RevisionStatusWidget.test.tsx:177]: I find the tests a little harder to follow when we're asserting a variable rather than the actual dropdown text, but no strong feelings

**Comment by Sepehr-Sobhani** [comment]
> Looking good! A couple overall notes:
> 
> * when I'm manually testing, when I change from Draft to anything except Applied, the pending actions from dropdown switches to readonly. Is that supposed to happen? I'd imagine users would still want to notify people up until something is Applied.
> * the form bounces around as the info text appears: https://www.loom.com/share/234d7037c9da4171aa89f05a514159ef

@BCerki 
Good call about the first point although It wasn't part of this ticket but I'm gonna fix it.
Regarding bounces, I think we need to make a separate ticket to capture that because our current margins and spacing don't support all those elements within the `DefaultLayout`. I've played with it a little bit changing the button and the select widget sizing but it didn't help. do you have any suggestions/fixes for that?

File: app/components/ProjectRevision/ChangeReasonWidget.tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    const newValue = e.target.value;
+
+    let valueChanged: boolean = initialValue !== newValue;
+    setUpdated(!valueChanged);
+    // set the text next to the button based on whether the value has changed or not
+    setInformationalText(
+      valueChanged
+        ? 'To confirm your change, please click the "Update" button.'
+        : ""
+    );
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/components/ProjectRevision/ChangeReasonWidget.tsx:67]: Great feedback

File: app/pages/cif/project-revision/[projectRevision]/view.tsx [hunk 6]
```diff
@@ -76,6 +78,8 @@ export const ViewProjectRevisionQuery = graphql`
       revisionType
       createdAt
       revisionStatus
+      # eslint-disable-next-line relay/unused-fields
+      pendingActionsFrom
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/pages/cif/project-revision/[projectRevision]/view.tsx:82]: It gets passed down to the widget via the form data.

File: app/components/ProjectRevision/ChangeReasonWidget.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+import { graphql, useFragment } from "react-relay";
+
+const ChangeReasonWidgetFragment = graphql`
+  fragment ChangeReasonWidget_projectRevision on ProjectRevision {
+    id
+    changeStatus
+  }
+`;
+
+// Custom widget to update the revision change reason
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/components/ProjectRevision/ChangeReasonWidget.tsx:14]: I agree that the naming is confusing, I had the same issue before in other fields since not all fields are named the same as their representative in the DB. I usually first look at the schema to understand the relationship between those.

File: app/components/ProjectRevision/ChangeReasonWidget.tsx [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+              style={{ marginRight: "1rem" }}
+              disabled={
+                initialValue === (value ?? "") || isUpdatingProjectRevision
+              }
+              size="small"
+            >
+              Update
+            </Button>
+          </div>
+          <small>{informationalText}</small>
```
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/components/ProjectRevision/ChangeReasonWidget.tsx:87]: I think we still need the `informationalText` state since we update it once `updateProjectRevision` is completed; we change its value to `Updated` once the mutation is done.
I can make `<small>` tag optional to not render the empty tag though.

File: app/pages/cif/project-revision/[projectRevision]/view.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Sepehr-Sobhani) [reply app/pages/cif/project-revision/[projectRevision]/view.tsx:82]: @BCerki I tried this approach but it makes the logic within the component hard to follow as it has its own state when changing the value from the select widget. I agree with keeping the data close to where it belongs but in this case, it makes the code more complicated to follow. If you still feel we can change the code to satisfy this approach, I would be more than happy to pair on this with you.

File: app/components/ProjectRevision/ChangeReasonWidget.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (BCerki) [reply app/components/ProjectRevision/ChangeReasonWidget.tsx:14]: Sounds like you already put a lot of thought into it, happy to leave it
